compilation error on arm64: undefined symbol deserializeAndInsertFromPosImpl

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
find an arm64 machine and build tiflash
```
[root@10 rel]# lscpu
架构：               aarch64
  CPU 运行模式：     64-bit
  字节序：           Little Endian
```

```
cmake ../.. \
      -GNinja \
      -DUSE_INTERNAL_SSL_LIBRARY=ON \
      -Wno-dev \
      -DNO_WERROR=ON

cmake --build . --target tiflash --parallel 30 &> t.err
```
<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)
build ok
### 3. What did you see instead (Required)
```
>>> referenced by FunctionsConversion.cpp
>>>               FunctionsConversion.cpp.o:(DB::ColumnDecimal<DB::Decimal<int>>::deserializeForCmpAndInsertFromPos(DB::PODArray<char*, 4096ul, Allocator<false>, 15ul, 16ul>&, bool)) in archive dbms/src/Functions/libtiflash_functions.a

ld.lld: error: undefined symbol: void DB::ColumnDecimal<DB::Decimal<int>>::deserializeAndInsertFromPosImpl<false>(DB::PODArray<char*, 4096ul, Allocator<false>, 15ul, 16ul>&, bool)
>>> referenced by FunctionsConversion.cpp
>>>               FunctionsConversion.cpp.o:(DB::ColumnDecimal<DB::Decimal<int>>::deserializeAndInsertFromPos(DB::PODArray<char*, 4096ul, Allocator<false>, 15ul, 16ul>&, bool)) in archive dbms/src/Functions/libtiflash_functions.a

ld.lld: error: undefined symbol: void DB::ColumnDecimal<DB::Decimal<long>>::deserializeAndInsertFromPosImpl<true>(DB::PODArray<char*, 4096ul, Allocator<false>, 15ul, 16ul>&, bool)
>>> referenced by FunctionsConversion.cpp
>>>               FunctionsConversion.cpp.o:(DB::ColumnDecimal<DB::Decimal<long>>::deserializeForCmpAndInsertFromPos(DB::PODArray<char*, 4096ul, Allocator<false>, 15ul, 16ul>&, bool)) in archive dbms/src/Functions/libtiflash_functions.a

ld.lld: error: undefined symbol: void DB::ColumnDecimal<DB::Decimal<long>>::deserializeAndInsertFromPosImpl<false>(DB::PODArray<char*, 4096ul, Allocator<false>, 15ul, 16ul>&, bool)
>>> referenced by FunctionsConversion.cpp
>>>               FunctionsConversion.cpp.o:(DB::ColumnDecimal<DB::Decimal<long>>::deserializeAndInsertFromPos(DB::PODArray<char*, 4096ul, Allocator<false>, 15ul, 16ul>&, bool)) in archive dbms/src/Functions/libtiflash_functions.a

ld.lld: error: undefined symbol: void DB::ColumnDecimal<DB::Decimal<__int128>>::deserializeAndInsertFromPosImpl<true>(DB::PODArray<char*, 4096ul, Allocator<false>, 15ul, 16ul>&, bool)
>>> referenced by FunctionsConversion.cpp
>>>               FunctionsConversion.cpp.o:(DB::ColumnDecimal<DB::Decimal<__int128>>::deserializeForCmpAndInsertFromPos(DB::PODArray<char*, 4096ul, Allocator<false>, 15ul, 16ul>&, bool)) in archive dbms/src/Functions/libtiflash_functions.a

ld.lld: error: undefined symbol: void DB::ColumnDecimal<DB::Decimal<__int128>>::deserializeAndInsertFromPosImpl<false>(DB::PODArray<char*, 4096ul, Allocator<false>, 15ul, 16ul>&, bool)
>>> referenced by FunctionsConversion.cpp
>>>               FunctionsConversion.cpp.o:(DB::ColumnDecimal<DB::Decimal<__int128>>::deserializeAndInsertFromPos(DB::PODArray<char*, 4096ul, Allocator<false>, 15ul, 16ul>&, bool)) in archive dbms/src/Functions/libtiflash_functions.a

ld.lld: error: undefined symbol: void DB::ColumnDecimal<DB::Decimal<boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)1, (boost::multiprecision::cpp_int_check_type)1, void>, (boost::multiprecision::expression_template_option)0>>>::deserializeAndInsertFromPosImpl<false>(DB::PODArray<char*, 4096ul, Allocator<false>, 15ul, 16ul>&, bool)
>>> referenced by FunctionsConversion.cpp
>>>               FunctionsConversion.cpp.o:(DB::ColumnDecimal<DB::Decimal<boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<256u, 256u, (boost::multiprecision::cpp_integer_type)1, (boost::multiprecision::cpp_int_check_type)1, void>, (boost::multiprecision::expression_template_option)0>>>::deserializeAndInsertFromPos(DB::PODArray<char*, 4096ul, Allocator<false>, 15ul, 16ul>&, bool)) in archive dbms/src/Functions/libtiflash_functions.a
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```
### 4. What is your TiFlash version? (Required)
master: 57122faa0fb94a330becd89369d0c6e59a1e051d

--- a/dbms/src/Columns/ColumnDecimal.cpp
+++ b/dbms/src/Columns/ColumnDecimal.cpp
@@ -140,6 +140,137 @@ const char * ColumnDecimal<T>::deserializeAndInsertFromArena(const char * pos, c
 }
 
 template <typename T>
+void ColumnDecimal<T>::countSerializeByteSizeForCmp(PaddedPODArray<size_t> & byte_size, const TiDB::TiDBCollatorPtr &) const
+{
+    countSerializeByteSizeImpl<true>(byte_size);
+}
+
+template <typename T>
+void ColumnDecimal<T>::countSerializeByteSize(PaddedPODArray<size_t> & byte_size) const
+{
+    countSerializeByteSizeImpl<false>(byte_size);
+}
+
+
+template <typename T>
+void ColumnDecimal<T>::countSerializeByteSizeForCmpColumnArray(
+        PaddedPODArray<size_t> & byte_size,
+        const IColumn::Offsets & array_offsets,
+        const TiDB::TiDBCollatorPtr &) const
+{
+    countSerializeByteSizeForColumnArrayImpl<true>(byte_size, array_offsets);
+}
+
+template <typename T>
+void ColumnDecimal<T>::countSerializeByteSizeForColumnArray(
+        PaddedPODArray<size_t> & byte_size,
+        const IColumn::Offsets & array_offsets) const
+{
+    countSerializeByteSizeForColumnArrayImpl<false>(byte_size, array_offsets);
+}
+
+
+template <typename T>
+void ColumnDecimal<T>::serializeToPosForCmp(
+        PaddedPODArray<char *> & pos,
+        size_t start,
+        size_t length,
+        bool has_null,
+        const TiDB::TiDBCollatorPtr &,
+        String *) const
+{
+    if (has_null)
+        serializeToPosImpl</*has_null=*/true, /*for_compare=*/true>(pos, start, length);
+    else
+        serializeToPosImpl</*has_null=*/false, /*for_compare=*/true>(pos, start, length);
+}
+
+template <typename T>
+void ColumnDecimal<T>::serializeToPos(PaddedPODArray<char *> & pos, size_t start, size_t length, bool has_null) const
+{
+    if (has_null)
+        serializeToPosImpl</*has_null=*/true, /*for_compare=*/false>(pos, start, length);
+    else
+        serializeToPosImpl</*has_null=*/false, /*for_compare=*/false>(pos, start, length);
+}
+
+template <typename T>
+void ColumnDecimal<T>::serializeToPosForCmpColumnArray(
+        PaddedPODArray<char *> & pos,
+        size_t start,
+        size_t length,
+        bool has_null,
+        const IColumn::Offsets & array_offsets,
+        const TiDB::TiDBCollatorPtr &,
+        String *) const
+{
+    if (has_null)
+        serializeToPosForColumnArrayImpl</*has_null=*/true, /*for_compare=*/true>(
+                pos,
+                start,
+                length,
+                array_offsets);
+    else
+        serializeToPosForColumnArrayImpl</*has_null=*/false, /*for_compare=*/true>(
+                pos,
+                start,
+                length,
+                array_offsets);
+}
+
+template <typename T>
+void ColumnDecimal<T>::serializeToPosForColumnArray(
+        PaddedPODArray<char *> & pos,
+        size_t start,
+        size_t length,
+        bool has_null,
+        const IColumn::Offsets & array_offsets) const
+{
+    if (has_null)
+        serializeToPosForColumnArrayImpl</*has_null=*/true, /*for_compare=*/false>(
+                pos,
+                start,
+                length,
+                array_offsets);
+    else
+        serializeToPosForColumnArrayImpl</*has_null=*/false, /*for_compare=*/false>(
+                pos,
+                start,
+                length,
+                array_offsets);
+}
+
+template <typename T>
+void ColumnDecimal<T>::deserializeForCmpAndInsertFromPos(PaddedPODArray<char *> & pos, bool use_nt_align_buffer)
+{
+    deserializeAndInsertFromPosImpl<true>(pos, use_nt_align_buffer);
+}
+
+template <typename T>
+void ColumnDecimal<T>::deserializeAndInsertFromPos(PaddedPODArray<char *> & pos, bool use_nt_align_buffer)
+{
+    deserializeAndInsertFromPosImpl<false>(pos, use_nt_align_buffer);
+}
+
+template <typename T>
+void ColumnDecimal<T>::deserializeForCmpAndInsertFromPosColumnArray(
+        PaddedPODArray<char *> & pos,
+        const IColumn::Offsets & array_offsets,
+        bool use_nt_align_buffer)
+{
+    deserializeAndInsertFromPosForColumnArrayImpl<true>(pos, array_offsets, use_nt_align_buffer);
+}
+
+template <typename T>
+void ColumnDecimal<T>::deserializeAndInsertFromPosForColumnArray(
+        PaddedPODArray<char *> & pos,
+        const IColumn::Offsets & array_offsets,
+        bool use_nt_align_buffer)
+{
+    deserializeAndInsertFromPosForColumnArrayImpl<false>(pos, array_offsets, use_nt_align_buffer);
+}
+
+template <typename T>
 template <bool for_compare>
 void ColumnDecimal<T>::countSerializeByteSizeImpl(PaddedPODArray<size_t> & byte_size) const
 {
@@ -160,7 +291,6 @@ void ColumnDecimal<T>::countSerializeByteSizeImpl(PaddedPODArray<size_t> & byte_
     }
 }
 
-// TODO add unit test
 template <typename T>
 template <bool for_compare>
 void ColumnDecimal<T>::countSerializeByteSizeForColumnArrayImpl(

--- a/dbms/src/Columns/ColumnDecimal.cpp
+++ b/dbms/src/Columns/ColumnDecimal.cpp
@@ -140,7 +140,8 @@ const char * ColumnDecimal<T>::deserializeAndInsertFromArena(const char * pos, c
 }
 
 template <typename T>
-void ColumnDecimal<T>::countSerializeByteSizeForCmp(PaddedPODArray<size_t> & byte_size, const TiDB::TiDBCollatorPtr &) const
+void ColumnDecimal<T>::countSerializeByteSizeForCmp(PaddedPODArray<size_t> & byte_size, const TiDB::TiDBCollatorPtr &)
+    const
 {
     countSerializeByteSizeImpl<true>(byte_size);
 }
@@ -154,17 +155,17 @@ void ColumnDecimal<T>::countSerializeByteSize(PaddedPODArray<size_t> & byte_size
 
 template <typename T>
 void ColumnDecimal<T>::countSerializeByteSizeForCmpColumnArray(
-        PaddedPODArray<size_t> & byte_size,
-        const IColumn::Offsets & array_offsets,
-        const TiDB::TiDBCollatorPtr &) const
+    PaddedPODArray<size_t> & byte_size,
+    const IColumn::Offsets & array_offsets,
+    const TiDB::TiDBCollatorPtr &) const
 {
     countSerializeByteSizeForColumnArrayImpl<true>(byte_size, array_offsets);
 }
 
 template <typename T>
 void ColumnDecimal<T>::countSerializeByteSizeForColumnArray(
-        PaddedPODArray<size_t> & byte_size,
-        const IColumn::Offsets & array_offsets) const
+    PaddedPODArray<size_t> & byte_size,
+    const IColumn::Offsets & array_offsets) const
 {
     countSerializeByteSizeForColumnArrayImpl<false>(byte_size, array_offsets);
 }
@@ -172,12 +173,12 @@ void ColumnDecimal<T>::countSerializeByteSizeForColumnArray(
 
 template <typename T>
 void ColumnDecimal<T>::serializeToPosForCmp(
-        PaddedPODArray<char *> & pos,
-        size_t start,
-        size_t length,
-        bool has_null,
-        const TiDB::TiDBCollatorPtr &,
-        String *) const
+    PaddedPODArray<char *> & pos,
+    size_t start,
+    size_t length,
+    bool has_null,
+    const TiDB::TiDBCollatorPtr &,
+    String *) const
 {
     if (has_null)
         serializeToPosImpl</*has_null=*/true, /*for_compare=*/true>(pos, start, length);
@@ -196,48 +197,32 @@ void ColumnDecimal<T>::serializeToPos(PaddedPODArray<char *> & pos, size_t start
 
 template <typename T>
 void ColumnDecimal<T>::serializeToPosForCmpColumnArray(
-        PaddedPODArray<char *> & pos,
-        size_t start,
-        size_t length,
-        bool has_null,
-        const IColumn::Offsets & array_offsets,
-        const TiDB::TiDBCollatorPtr &,
-        String *) const
+    PaddedPODArray<char *> & pos,
+    size_t start,
+    size_t length,
+    bool has_null,
+    const IColumn::Offsets & array_offsets,
+    const TiDB::TiDBCollatorPtr &,
+    String *) const
 {
     if (has_null)
-        serializeToPosForColumnArrayImpl</*has_null=*/true, /*for_compare=*/true>(
-                pos,
-                start,
-                length,
-                array_offsets);
+        serializeToPosForColumnArrayImpl</*has_null=*/true, /*for_compare=*/true>(pos, start, length, array_offsets);
     else
-        serializeToPosForColumnArrayImpl</*has_null=*/false, /*for_compare=*/true>(
-                pos,
-                start,
-                length,
-                array_offsets);
+        serializeToPosForColumnArrayImpl</*has_null=*/false, /*for_compare=*/true>(pos, start, length, array_offsets);
 }
 
 template <typename T>
 void ColumnDecimal<T>::serializeToPosForColumnArray(
-        PaddedPODArray<char *> & pos,
-        size_t start,
-        size_t length,
-        bool has_null,
-        const IColumn::Offsets & array_offsets) const
+    PaddedPODArray<char *> & pos,
+    size_t start,
+    size_t length,
+    bool has_null,
+    const IColumn::Offsets & array_offsets) const
 {
     if (has_null)
-        serializeToPosForColumnArrayImpl</*has_null=*/true, /*for_compare=*/false>(
-                pos,
-                start,
-                length,
-                array_offsets);
+        serializeToPosForColumnArrayImpl</*has_null=*/true, /*for_compare=*/false>(pos, start, length, array_offsets);
     else
-        serializeToPosForColumnArrayImpl</*has_null=*/false, /*for_compare=*/false>(
-                pos,
-                start,
-                length,
-                array_offsets);
+        serializeToPosForColumnArrayImpl</*has_null=*/false, /*for_compare=*/false>(pos, start, length, array_offsets);
 }
 
 template <typename T>
@@ -254,18 +239,18 @@ void ColumnDecimal<T>::deserializeAndInsertFromPos(PaddedPODArray<char *> & pos,
 
 template <typename T>
 void ColumnDecimal<T>::deserializeForCmpAndInsertFromPosColumnArray(
-        PaddedPODArray<char *> & pos,
-        const IColumn::Offsets & array_offsets,
-        bool use_nt_align_buffer)
+    PaddedPODArray<char *> & pos,
+    const IColumn::Offsets & array_offsets,
+    bool use_nt_align_buffer)
 {
     deserializeAndInsertFromPosForColumnArrayImpl<true>(pos, array_offsets, use_nt_align_buffer);
 }
 
 template <typename T>
 void ColumnDecimal<T>::deserializeAndInsertFromPosForColumnArray(
-        PaddedPODArray<char *> & pos,
-        const IColumn::Offsets & array_offsets,
-        bool use_nt_align_buffer)
+    PaddedPODArray<char *> & pos,
+    const IColumn::Offsets & array_offsets,
+    bool use_nt_align_buffer)
 {
     deserializeAndInsertFromPosForColumnArrayImpl<false>(pos, array_offsets, use_nt_align_buffer);
 }

--- a/dbms/src/Columns/ColumnDecimal.h
+++ b/dbms/src/Columns/ColumnDecimal.h
@@ -173,28 +173,16 @@ public:
         String &) const override;
     const char * deserializeAndInsertFromArena(const char * pos, const TiDB::TiDBCollatorPtr &) override;
 
-    void countSerializeByteSizeForCmp(PaddedPODArray<size_t> & byte_size, const TiDB::TiDBCollatorPtr &) const override
-    {
-        countSerializeByteSizeImpl<true>(byte_size);
-    }
-    void countSerializeByteSize(PaddedPODArray<size_t> & byte_size) const override
-    {
-        countSerializeByteSizeImpl<false>(byte_size);
-    }
+    void countSerializeByteSizeForCmp(PaddedPODArray<size_t> & byte_size, const TiDB::TiDBCollatorPtr &) const override;
+    void countSerializeByteSize(PaddedPODArray<size_t> & byte_size) const override;
 
     void countSerializeByteSizeForCmpColumnArray(
         PaddedPODArray<size_t> & byte_size,
         const IColumn::Offsets & array_offsets,
-        const TiDB::TiDBCollatorPtr &) const override
-    {
-        countSerializeByteSizeForColumnArrayImpl<true>(byte_size, array_offsets);
-    }
+        const TiDB::TiDBCollatorPtr &) const override;
     void countSerializeByteSizeForColumnArray(
         PaddedPODArray<size_t> & byte_size,
-        const IColumn::Offsets & array_offsets) const override
-    {
-        countSerializeByteSizeForColumnArrayImpl<false>(byte_size, array_offsets);
-    }
+        const IColumn::Offsets & array_offsets) const override;
 
     void serializeToPosForCmp(
         PaddedPODArray<char *> & pos,
@@ -202,20 +190,8 @@ public:
         size_t length,
         bool has_null,
         const TiDB::TiDBCollatorPtr &,
-        String *) const override
-    {
-        if (has_null)
-            serializeToPosImpl</*has_null=*/true, /*for_compare=*/true>(pos, start, length);
-        else
-            serializeToPosImpl</*has_null=*/false, /*for_compare=*/true>(pos, start, length);
-    }
-    void serializeToPos(PaddedPODArray<char *> & pos, size_t start, size_t length, bool has_null) const override
-    {
-        if (has_null)
-            serializeToPosImpl</*has_null=*/true, /*for_compare=*/false>(pos, start, length);
-        else
-            serializeToPosImpl</*has_null=*/false, /*for_compare=*/false>(pos, start, length);
-    }
+        String *) const override;
+    void serializeToPos(PaddedPODArray<char *> & pos, size_t start, size_t length, bool has_null) const override;
 
     void serializeToPosForCmpColumnArray(
         PaddedPODArray<char *> & pos,
@@ -224,65 +200,25 @@ public:
         bool has_null,
         const IColumn::Offsets & array_offsets,
         const TiDB::TiDBCollatorPtr &,
-        String *) const override
-    {
-        if (has_null)
-            serializeToPosForColumnArrayImpl</*has_null=*/true, /*for_compare=*/true>(
-                pos,
-                start,
-                length,
-                array_offsets);
-        else
-            serializeToPosForColumnArrayImpl</*has_null=*/false, /*for_compare=*/true>(
-                pos,
-                start,
-                length,
-                array_offsets);
-    }
+        String *) const override;
     void serializeToPosForColumnArray(
         PaddedPODArray<char *> & pos,
         size_t start,
         size_t length,
         bool has_null,
-        const IColumn::Offsets & array_offsets) const override
-    {
-        if (has_null)
-            serializeToPosForColumnArrayImpl</*has_null=*/true, /*for_compare=*/false>(
-                pos,
-                start,
-                length,
-                array_offsets);
-        else
-            serializeToPosForColumnArrayImpl</*has_null=*/false, /*for_compare=*/false>(
-                pos,
-                start,
-                length,
-                array_offsets);
-    }
+        const IColumn::Offsets & array_offsets) const override;
 
-    void deserializeForCmpAndInsertFromPos(PaddedPODArray<char *> & pos, bool use_nt_align_buffer) override
-    {
-        deserializeAndInsertFromPosImpl<true>(pos, use_nt_align_buffer);
-    }
-    void deserializeAndInsertFromPos(PaddedPODArray<char *> & pos, bool use_nt_align_buffer) override
-    {
-        deserializeAndInsertFromPosImpl<false>(pos, use_nt_align_buffer);
-    }
+    void deserializeForCmpAndInsertFromPos(PaddedPODArray<char *> & pos, bool use_nt_align_buffer) override;
+    void deserializeAndInsertFromPos(PaddedPODArray<char *> & pos, bool use_nt_align_buffer) override;
 
     void deserializeForCmpAndInsertFromPosColumnArray(
         PaddedPODArray<char *> & pos,
         const IColumn::Offsets & array_offsets,
-        bool use_nt_align_buffer) override
-    {
-        deserializeAndInsertFromPosForColumnArrayImpl<true>(pos, array_offsets, use_nt_align_buffer);
-    }
+        bool use_nt_align_buffer) override;
     void deserializeAndInsertFromPosForColumnArray(
         PaddedPODArray<char *> & pos,
         const IColumn::Offsets & array_offsets,
-        bool use_nt_align_buffer) override
-    {
-        deserializeAndInsertFromPosForColumnArrayImpl<false>(pos, array_offsets, use_nt_align_buffer);
-    }
+        bool use_nt_align_buffer) override;
 
     void flushNTAlignBuffer() override;
 

